apis/python/src/tiledbsoma/fastercsx.cc

```diff
@@ -0,0 +1,621 @@
+/**
+ * @file   fastercsx.cc
+ *
+ * @section LICENSE
+ *
+ * The MIT License
+ *
+ * @copyright Copyright (c) 2024 TileDB, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * Python bindings for CSX conversion primitives.
+ */
+
+#include <variant>
+
+// Define to include extra debugging bindings (e.g., count_rows)
+/* #define FASTERCSX__DEBUGGING_HOOKS 1 */
+
+#include <tiledbsoma/utils/fastercsx.h>
+#include "common.h"
+
+namespace libtiledbsomacpp {
+
+namespace py = pybind11;
+using npy_api = pybind11::detail::npy_api;
+
+using namespace py::literals;
+using namespace tiledbsoma;
+
+/**
+ * @brief Convert py::tuple<py::array> to std::vector<py::array>.
+ *
+ * Will throw if the tuple does not contain a Numpy ndarray of the correct type.
+ *
+ * @tparam T
+ * @param tup
+ * @return std::vector<T>
+ */
+template <typename T>
+std::vector<T> to_vector_(const py::tuple& tup) {
+    auto vec = std::vector<T>();
+    vec.reserve(tup.size());
+    for (py::size_t i = 0; i < tup.size(); ++i) {
+        auto v = tup[i];
+        if (!py::isinstance<T>(v))
+            throw py::type_error(
+                std::string("Unable to cast tuple element to '") +
+                py::type_id<T>() + "'");
+        vec.push_back(v.cast<T>());
+    }
+    return vec;
+}
+
+/**
+ * @brief Various helpers which convert from py:array to spans.
+ */
+
+template <typename T>
+tcb::span<T const> make_span_(py::array arr) {
```

Comment:
Can this be `const py::array& arr`? (Likewise with functions after this)

Reply:
To what end? 

I believe it is simply a typed pointer, and all of the PyBind docs/example code use the above pattern.  I'd rather stick with the recommended pattern unless there is a benefit I'm unaware of...  Thoughts?

Reply:
Thanks for the feedback that it's a typed pointer -- I did not grok that from my read -- thanks for the info!